File: 01000/1260 - DFS와 BFS.py
__problem__ = 'https://boj.kr/1260', 'DFS와 BFS'

import sys
from collections import deque

input = sys.stdin.readline

n, m, v = map(int, input().split())
graph = [[] for _ in range(n+1)]

for _ in range(m):
    a, b = map(int, input().split())

    graph[a].append(b)
    graph[b].append(a)

# 작은 것부터 방문해야 하므로 sort 해야됨
for i in range(n+1):
    graph[i].sort()

def dfs(graph, now, visited):
    visited[now] = True
    print(now, end=' ')

    for w in graph[now]:
        if not visited[w]:
            dfs(graph, w, visited)

def bfs(graph, start):
    visited = [False]*(n+1)
    visited[start] = True
    queue = deque([start])

    while queue:
        top = queue.popleft()
        print(top, end=' ')

        for w in graph[top]:
            if not visited[w]:
                queue.append(w)
                visited[w] = True

dfs(graph, v, [False]*(n+1))
print()
bfs(graph, v)

File: 01000/1260 - DFS와 BFS.cpp
#include <bits/stdc++.h>
#include <ranges>
#define endl "\n"
#define INF 0x3f3f3f3f

using namespace std;

void dfs(const vector<vector<int>> &adj, const int start) {
    stack<int> stack;
    unordered_set<int> visited;

    stack.push(start);

    while (!stack.empty()) {
        const int vtx = stack.top();
        stack.pop();

        if (visited.contains(vtx)) continue;
        visited.insert(vtx);
        cout << vtx << " ";

        for (auto u : ranges::reverse_view(adj[vtx])) {
            if (!visited.contains(u)) {
                stack.push(u);
            }
        }
    }
}

void bfs(const vector<vector<int>> &adj, const int start) {
    queue<int> queue;
    unordered_set<int> visited;

    queue.push(start);

    while (!queue.empty()) {
        const int vtx = queue.front();
        queue.pop();

        if (visited.contains(vtx)) continue;
        visited.insert(vtx);
        cout << vtx << " ";

        for (auto u : adj[vtx]) {
            if (!visited.contains(u)) {
                queue.push(u);
            }
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    int n, m, v;
    cin >> n >> m >> v;

    vector<vector<int>> adj(n + 1);

    for (int i = 0; i < m; ++i) {
        int a, b;
        cin >> a >> b;

        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    for (int i = 1; i < adj.size(); ++i) {
        ranges::sort(adj[i]);
    }

    dfs(adj, v);
    cout << endl;
    bfs(adj, v);

    return 0;
}